### PR TITLE
Bugfix conversion of single activities in pipeline json files

### PR DESCRIPTION
**Pull request by Alex-Swann**
Bugfix conversion of single activities in pipeline json files when handling dynamic variables.

If a pipeline has a variable that needs to be dynamic to accommodate this being utilised in different deployment environments (DEV, UAT, PROD), e.g.

```json
{
	"name": "Run all notebooks",
	"properties": {
		"activities": [
			{
				"name": "ForEach Notebook",
				"type": "ForEach",
				"dependsOn": [],
				"userProperties": [],
				"typeProperties": {
					"items": {
						"value": "<an_array_of_string_values_exposed_via_@item()>",
						"type": "Expression"
					},
					"activities": [
						{
							"name": "Child - Pipeline",
							"type": "ExecutePipeline",
							"dependsOn": [],
							"policy": {
								"secureInput": false
							},
							"userProperties": [],
							"typeProperties": {
								"pipeline": {
									"referenceName": "Child - Pipeline",
									"type": "PipelineReference"
								},
								"waitOnCompletion": true,
								"parameters": {
									"StorageUrl": {
										"value": "@variables('StorageUrl')",
										"type": "Expression"
									}
								}
							}
						}
					]
				}
			}
		],
		"variables": {
			"StorageUrl": {
				"type": "String",
				"defaultValue": "test-url"
			}
		},
		"folder": {
			"name": "Core Code"
		},
		"annotations": []
	}
}
```
the serialisation and deserialisation process in `Save-SynapseObjectAsFile.ps1` means the activities array, and any nested activities, are accidentally converted back into a object as opposed to an array of objects.

```json
{
	"name": "Run all notebooks",
	"properties": {
		"activities": {
		          "name": "ForEach Notebook",
		          "type": "ForEach",
		          "dependsOn": [],
		          "userProperties": [],
		          "typeProperties": {
			  "items": {
				  "value": "<an_array_of_string_values_exposed_via_@item()>",
				  "type": "Expression"
			  },
			"activities": {
			        "name": "Child - Pipeline",
			        "type": "ExecutePipeline",
			        "dependsOn": [],
			        "policy": {
				        "secureInput": false
			        },
			        "userProperties": [],
			        "typeProperties": {
				        "pipeline": {
					        "referenceName": "Child - Pipeline",
					        "type": "PipelineReference"
				        },
				        "waitOnCompletion": true,
				        "parameters": {
					"StorageUrl": {
						"value": "@variables('StorageUrl')",
						    "type": "Expression"
					 }
				    }
			       }
		          }
		     }
		},
		"variables": {
			"StorageUrl": {
				"type": "String",
				"defaultValue": "test-url"
			}
		},
		"folder": {
			"name": "Core Code"
		},
		"annotations": []
	}
}
```

This fails Synapse's REST-API validations when attempting to deploy. This PR is to start the conversation around addressing it (or fix it altogether).

### CODE REVIEW

**Comment by NowinskiK** [comment]
Thank you, Alex for your contribution. Could you please add a unit test to cover the use case?

**Comment by Alex-Swann** [comment]
> Thank you, Alex for your contribution. Could you please add a unit test to cover the use case?

I hadn't spotted those. Thanks, I spotted an edge case too. Let me know if they align with both the fix and everything else the PS file is responsible for 👍

File: private/Get-RequestHeader.ps1 [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-        $token = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
-        [System.Runtime.InteropServices.Marshal]::ZeroFreeBSTR($BSTR)
+    if ($SynapseToken.Token -is [System.Security.SecureString]) {
+        $ptr = [System.Runtime.InteropServices.Marshal]::SecureStringToCoTaskMemUnicode($SynapseToken.Token)
+        $token = [System.Runtime.InteropServices.Marshal]::PtrToStringUni($ptr)
+        [System.Runtime.InteropServices.Marshal]::ZeroFreeCoTaskMemUnicode($ptr)
     }
     else {
-        $token = $SynapseToken.token
+        $token = $SynapseToken.Token ?? $SynapseToken.token
```
Comment (NowinskiK) [inline private/Get-RequestHeader.ps1:9]: Avoid using operators supported in PS 7.x+ only.

**Review by NowinskiK — CHANGES_REQUESTED** [review]
